internal/transport/grpchttp2/framer.go

```diff
@@ -0,0 +1,311 @@
+
+func (f *PingFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *PingFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// GoAwayFrame is the representation of a [GOAWAY Frame]. It implements the
+// Frame interface. The GOAWAY frame is used to initiate shutdown of a
+// connection or to signal serious error conditions.
+//
+// [GOAWAY Frame]: https://httpwg.org/specs/rfc7540.html#GOAWAY
+type GoAwayFrame struct {
+	hdr          *FrameHeader
+	free         func()
+	LastStreamID uint32
+	Code         ErrCode
+	DebugData    []byte
+}
+
+func (f *GoAwayFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *GoAwayFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// WindowUpdateFrame is the representation of a [WINDOW_UPDATE Frame]. It
+// implements the Frame interface. The WINDOW_UPDATE frame is used to implement
+// flow control.
+//
+// [WINDOW_UPDATE Frame]: https://httpwg.org/specs/rfc7540.html#WINDOW_UPDATE
+type WindowUpdateFrame struct {
+	hdr *FrameHeader
+	Inc uint32
+}
+
+func (f *WindowUpdateFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+// ContinuationFrame is the representation of a [CONTINUATION Frame]. It
+// implements the Frame interface. The CONTINUATION frame (type=0x9) is used to
+// continue a sequence of header block fragments.
+//
+// [CONTINUATION Frame]: https://httpwg.org/specs/rfc7540.html#CONTINUATION
+type ContinuationFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *ContinuationFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *ContinuationFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// MetaHeadersFrame is the representation of one HEADERS frame and zero or more
+// contiguous CONTINUATION frames and the decoding of their HPACK-encoded
+// contents.  This frame type is not transmitted over the network and is only
+// generated by the ReadFrame() function.
+type MetaHeadersFrame struct {
+	hdr    *FrameHeader
+	Fields []hpack.HeaderField
+	// Truncated indicates whether the MetaHeadersFrame has been truncated due
+	// to being longer than the MaxHeaderListSize.
+	Truncated bool
+}
+
+func (f *MetaHeadersFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *MetaHeadersFrame) Free() {}
+
+// Framer encapsulates the functionality to read and write HTTP/2 frames.
+type Framer interface {
+	// SetMetaDecoder will set a decoder for the framer. When the decoder is
+	// set, ReadFrame will parse the header values, merging all Headers and
+	// Continuation frames.
+	SetMetaDecoder(d *hpack.Decoder)
+	// ReadFrame returns grpchttp2.Frame. It is the caller's responsibility to
+	// call Frame.Free() once it is done using it.
+	ReadFrame() (Frame, error)
+	// WriteData writes an HTTP/2 DATA frame to the stream.
+	WriteData(streamID uint32, endStream bool, data ...[]byte) error
+	// WriteData writes an HTTP/2 HEADERS frame to the stream.
+	WriteHeaders(streamID uint32, endStream, endHeaders bool, headerBlock ...[]byte) error
```

Comment:
It would be nice to add a TODO either to the docstring of this interface (mentioning the methods which would be affected) or to every affected method about the impending change to `mem.Buffer` from `[]byte`

Reply:
Done.